docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,201 @@
+
+* `--eviction-strategy=:"` - The strategy to use when evicting pods
+when eviction criteria have been met.
+
+The `kubelet` will target `BestEffort`, `Burstable`, and `Guaranteed`
+pods in that order to find a candidate set of pods to fail to force reclamation
+of the starved compute resource.  Within each quality of service tier,
+the `kubelet` will evaluate each pod relative to the eviction strategy.
+
+If the `kubelet` fails a pod based on eviction criteria, an `Event` is created
+describing why the `Pod` was failed by the `kubelet`.
+
+This proposal describes the `greedy` strategy as the first candidate to
+implement based on the ability for operators to easily reason about it.
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
+starved compute resource relative to their scheduling guarantee.
+
+It ranks pods within a quality of service tier as follows:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+### Other potential eviction strategies
+
+It is anticipated other eviction strategies may be used in the future.
+
+For example, operators may want to target `BestEffort` pods that were most
+recently scheduled on the node instead of targeting `BestEffort` pods that
+consumed the most resource.
+
+Another potential option is that operators may want to target `BestEffort` pods
+that consume the least amount of resource assuming their sum would be enough
+to reclaim enough resource on the node.  The idea being that those low resource
+pods could fit elsewhere on the cluster.
+
+Finally, operators may want to target pods based on other criteria altogether.
+For example, the time it takes for a pod to be ready and accept traffic may be
+a valid concern when looking at which pods to target for failure since many of
+those pods may get rescheduled and restart on other nodes with significant delay.
+
+Given the number of potential strategies that could be utilized, this proposal
+anticipates other strategies being defined in the future based on more specialized
+operator use cases.
+
+## Improve node stability for BestEffort pods
+
+### Feasibility checks during kubelet admission
+
+The `kubelet` will reject `BestEffort` pods if the current availability is
+below the configured eviction-threshold independent of the configured
+eviction window.
+
+The `kubelet` supports a flag to enable this behavior:
+
+* `--eviction-on-admission=true` - Do not start and fail BestEffort pods if 
```

Comment:
I don't intend to ever run with the flag set to false, but put a flag there since I thought @davidopp expressed some reservation on the Kubelet doing feasibility checks until the scheduler had more back off capability.

The other reason would be if the operator did not intend to evict on most recently scheduled since this basically gives guaranteed lowest priority to that class of best effort pod.